lms/djangoapps/courseware/views.py

```diff
@@ -1419,3 +1426,180 @@ def render_xblock(request, usage_key_string, check_if_enrolled=True):
+    except InvalidRequestError:
+        return HttpResponseServerError(error_message)
+    except TimeoutError:
+        return HttpResponseServerError(error_message)
+
+    return process_analytics_answer_dist(data, question_types_by_part, num_options_by_part)
+
+
+def process_analytics_answer_dist(data, question_types_by_part, num_options_by_part):
+    """
+    Aggregates the analytics answer dist data.
+    From the data, gets the date/time the data was last updated and reformats to the client TZ.
+
+    Arguments:
+        data: response from the analytics api
+        question_types_by_part: dict of question types
+        num_options_by_part: dict of number of options by question
+
+    Returns:
+        A json payload of:
+          - data by part: an array of dicts of {value_id, correct, count} for each part_id
+          - count by part: an array of dicts of {totalAttemptCount, totalCorrectCount, TotalIncorrectCount} for each part_id
+          - last updated: string
+     """
+
+    # Each element in count_by_part is a dict of totalAttemptCount, totalCorrectCount, totalIncorrectCount
+    count_by_part = {}
+
+    # Each element in data_by_part is an array of dicts of {value_id, correct, count}
+    data_by_part = {}
+
+    # For errors discovered during analytics data processing use message_by_part
+    message_by_part = {}
+
+    # Count rows returned for each part for integrity check
+    num_rows_by_part = {}
+
+    for item in data:
+        part_id = item['part_id']
+
+        num_rows_by_part[part_id] = num_rows_by_part.get(part_id, 0) + 1
+
+        # If we detect an issue with the data, set error message and continue
+        if _issue_with_data(item,
+                            part_id,
+                            message_by_part,
+                            question_types_by_part,
+                            num_options_by_part,
+                            num_rows_by_part):
+            continue
+
+        # Add count to appropriate aggregates
+        count_dict = count_by_part.get(part_id, {
+            'totalAttemptCount': 0,
+            'totalCorrectCount': 0,
+            'totalIncorrectCount': 0,
+        })
+        count_dict['totalAttemptCount'] = count_dict.get('totalAttemptCount') + item['count']
+        if item['correct']:
+            count_dict['totalCorrectCount'] = count_dict.get('totalCorrectCount') + item['count']
+        else:
+            count_dict['totalIncorrectCount'] = count_dict.get('totalIncorrectCount') + item['count']
+
+        count_by_part[part_id] = count_dict
+
+        # Add this item's data to the data for this part_id
+        part_dict = {}
+        part_dict['value_id'] = item['value_id']
+        part_dict['correct'] = item['correct']
+        part_dict['count'] = item['count']
+
+        data_by_part[part_id] = data_by_part.get(part_id, []) + [part_dict]
+
+    # Determine the last updated date, convert to client TZ and format
```

Comment:
Do we do this elsewhere? If not, I'd be worried about creating confusion, where some system dates are in UTC and the others are coerced. Honestly I don't remember what we do